src/basic/process-util.h

```diff
@@ -179,3 +179,18 @@ static inline int safe_fork(const char *name, ForkFlags flags, pid_t *ret_pid) {
 }
 
 int fork_agent(const char *name, const int except[], unsigned n_except, pid_t *pid, const char *path, ...);
+
+#if SIZEOF_PID_T == 4
+#define PID_T_MAX ((pid_t) INT32_MAX)
+/* The maximum number of PIDs Linux allows on this architecture. Tthis reflects a limit compiled into the kernel, and
+ * sets the limit on what may be written to the /proc/sys/kernel/pid_max sysctl (but do note that the sysctl is off by
+ * 1, since PID 0 can never exist). */
```

Comment:
This comment is misleading. The name "_max" is wrong because it is one above the allowed range (contrary to the definition of maximum, which in case of integers corresponds to the biggest allowed number), and the fact that the range is cut off at the bottom does not matter.

Reply:
hmm, no in this case the macro is actually both: the largest value and the total number of allowed items. that's because the range of pids is actually 1..n, i.e. because the counting starts at 1 (as that's the first valid PID), the highest PID is 4194303 and that's also the maximum number of concurrent tasks!

Reply:
also, i am not sure i agree with the rule that "_max" is strictly to be understood as "the highest valid value"... For sizing stuff it usually doesn't mean that, but the maximum number of objects, and if those objects are numbered from 0 this hence differs...

I mean, look at this list from POSIX:

http://pubs.opengroup.org/onlinepubs/009695399/basedefs/limits.h.html

More often than not in this list _MAX is used as the limit, not as the "highest possible value"

Reply:
particularly nice on that posix link are the items for NAME_MAX and PATH_MAX btw, because one includes the NUL byte, the other does not.

Reply:
(but nothing of this matters in this specific case as the TASKS_MAX definition here is *both* the limit and the highest possible value, as counting starts at 1 here, as mentioned)